### DISCUSSION TITLE
authToken security

### DISCUSSION THREAD

**Discussion by LarsSven** [gravitational/teleport#29805] [Q&A] [answered]
Hello everyone,

I need help with something. For the authToken that one adds to the helm chart's values.yml, what are the security implications of having it plaintext?

Obviously, it is best to obscure any tokens and not have them in the repository hosting the Kubernetes config files, however, I have seen mixed implementations, some hosting the token as a Kubernetes secret, and some putting them plaintext into the configuration file. While I will personally always opt for the secret, I need to know the security consequences of having this token be plaintext. Is there any or is such a token unusable to someone reading it?

I have not been able to find a consistent answer to this question yet.

Thank you in advance for your response.

**Comment by webvictim** [comment, marked as answer]
When joining Kubernetes agents to a cluster using a regular plain join token (for example, the output of `tctl tokens add --type=kube --ttl=1h`) then that particular token is valid for an infinite number of join attempts within the TTL. As such, anyone who gets hold of the token could potentially join their own Kubernetes agent to your cluster.

You can see the TTL on all currently valid tokens by running `tctl tokens ls` against your cluster.

As such, it's worth observing the following behaviours for security:
- Generate plain tokens for a short a period of time as possible (the shorter the TTL, the more secure the token is)
- Remove tokens prior to their expiry if they are no longer needed (`tctl tokens rm`)
- Don't publish or commit tokens in plaintext - use secrets to propagate them
- Use a non-token based join method if possible, like:
  - [AWS IAM](https://goteleport.com/docs/agents/join-services-to-your-cluster/aws-iam/)
  - [GCP Identity](https://goteleport.com/docs/agents/join-services-to-your-cluster/gcp/)
  - [Azure Identity](https://goteleport.com/docs/agents/join-services-to-your-cluster/azure/)
  - [Kubernetes identity](https://goteleport.com/docs/agents/join-services-to-your-cluster/kubernetes/).
